### PR TITLE
INT B-21657 - Prime request service items ppm

### PR DESCRIPTION
**Pull request by MInthavongsay**
## Summary

This PR introduces another validation layer to the Prime `operationId: createMTOServiceItem` API (version 1) to exclude operation for PPM shipments.  Service response will return 422 code indicating shipment type PPM is not allowable.

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

Note: **`operationId: createMTOServiceItem` API  is only at version 1.** Changes were also applied to version 2 and version 3 handlers to ensure the changes are cascaded to all potential API versions.

1. Login as customer and onboard/submit a move containing both HHG and PPM. For each use postal code **62226** for all pickup addresses and **02542** for all destination addresses. This is to ensure the entire workflow will be associated to the AGFM GBLOC.
2. Login as New Service Counselor for AGFM. Find move and submit move details.
3. Login as TOO for AGFM. Select HHG and Approve selected to add service items.
4. Login as Prime Sim for AGFM. Find move and Add Service Item for PPM. Verify ALL service item types are prevented from executing. Make sure to open web developer tool and capture JSON payload. Service Response must return 422 --  Error: Unprocessable Entity.
Screenshot 2024-12-17 at 11 07 07 AM
5. Goto http://primelocal:3000/swagger-ui/prime.html#/mtoServiceItem/createMTOServiceItem . Copy and paste JSON payload from step 4 and execute API. Verify response is 422:
 Screenshot 2024-12-17 at 11 08 04 AM
6. Verify shipment types NOT PPM can still create service items from Prime via Prime Sim, Swagger, Postman ..etc...

 

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

### CODE REVIEW

File: pkg/handlers/primeapiv2/mto_service_item.go [hunk 1]
```diff
@@ -42,6 +42,26 @@ func (h CreateMTOServiceItemHandler) Handle(params mtoserviceitemops.CreateMTOSe
 	return h.AuditableAppContextFromRequestWithErrors(params.HTTPRequest,
```
Comment (MInthavongsay) [inline pkg/handlers/primeapiv2/mto_service_item.go:42]: Note: Changes were also applied to version 2 and 3 handler.s Why was this added in the first place? It's currently not publicly exposed.

**Comment by CoryKleinjanCACI** [comment]
Should we remove the add service item button for ppm shipments from the prime sim now that they aren't allowed?

Screenshot 2024-12-17 at 12 14 28 PM

**Comment by MInthavongsay** [comment]
> Should we remove the add service item button for ppm shipments from the prime sim now that they aren't allowed?
> 
> Screenshot 2024-12-17 at 12 14 28 PM

I decided to leave the Add Service Item button in placed so we can have a quick and easy way to test the validation for PPM. We can in the future. For now, it might be a good reminder that PPMs are not permitted via Prime.

File: pkg/handlers/primeapi/mto_service_item.go [hunk 2]
```diff
@@ -44,6 +44,26 @@ func (h CreateMTOServiceItemHandler) Handle(params mtoserviceitemops.CreateMTOSe
...
+				}
+			}
+
+			if shipment.ShipmentType == models.MTOShipmentTypePPM {
+				verrs := validate.NewErrors()
+				verrs.Add("ShipmentType", string(shipment.ShipmentType))
+				appCtx.Logger().Error("primeapi.CreateMTOServiceItemHandler error", zap.Error(verrs))
+				return mtoserviceitemops.NewCreateMTOServiceItemUnprocessableEntity().WithPayload(payloads.ValidationError(
+					"Create Service Item is not allowed", h.GetTraceIDFromRequest(params.HTTPRequest), verrs)), verrs
+			}
```
Comment (antgmann) [inline pkg/handlers/primeapi/mto_service_item.go:65]: If the anticipated use case is the Prime attempting to create service items for a PPM when they shouldn't, would the error be more helpful to read "Create Service Item is not allowed for PPM Shipments" ?

**Review by antgmann — APPROVED** [review]
Personally I'm on the side of removing the button discussed above, but I won't contest it further. Just one small suggestion for the error we return, everything else seems good to me.

File: pkg/handlers/primeapi/mto_service_item.go (hunk 2 above)
Comment (antgmann) [reply pkg/handlers/primeapi/mto_service_item.go:65]: *this is present in multiple locations

Comment (MInthavongsay) [reply pkg/handlers/primeapi/mto_service_item.go:65]: Doesn't the error message convey that? It does say shipmentType PPM...

Comment (antgmann) [reply pkg/handlers/primeapi/mto_service_item.go:65]: It does say that there was an error due to it being a PPM Shipment, but it may not be immediately explicitly clear to the prime user that PPMs are not authorized to receive service items.

Comment (r-mettler) [reply pkg/handlers/primeapi/mto_service_item.go:65]: I think if you just change to "Create Service Items not allowed for:" then the rest would be more clear that its not allowed for PPM shipments.

**Comment by r-mettler** [comment]
image

Tests passed.

**Review by r-mettler — APPROVED** [review]

File: pkg/handlers/primeapi/mto_service_item.go (hunk 2 above)
Comment (MInthavongsay) [reply pkg/handlers/primeapi/mto_service_item.go:65]: I've updated the message. It will also include the mtoShipmentID for reference.

Screenshot 2024-12-18 at 1 23 08 PM

**Review by antgmann — APPROVED** [review]
Thanks!

**Review by r-mettler — APPROVED** [review]
